### DISCUSSION TITLE
Add ssh key error : 400

### DISCUSSION THREAD

**Discussion by koffiisen** [semaphoreui/semaphore#1861] [Q&A]
Impossible to add new ssh key

**_Semaphore v2.9.45_**

![image](https://github.com/ansible-semaphore/semaphore/assets/33517386/e8a61483-1b4d-494a-9436-ab3d53c489da)

![image](https://github.com/ansible-semaphore/semaphore/assets/33517386/0f2e1aad-16a3-4f7f-8fb4-fef1889b6e19)

```
ansible_ui   | time="2024-03-06T15:08:31Z" level=error msg="illegal base64 data at input byte 57"
ansible_ui   | goroutine 15 [running]:
ansible_ui   | runtime/debug.Stack()
ansible_ui   |  /usr/local/go/src/runtime/debug/stack.go:24 +0x65
ansible_ui   | runtime/debug.PrintStack()
ansible_ui   |  /usr/local/go/src/runtime/debug/stack.go:16 +0x19
ansible_ui   | github.com/ansible-semaphore/semaphore/api/helpers.WriteError({0x2345ac0, 0xc000510000}, {0x2341600?, 0x2778a08?})
ansible_ui   |  /go/src/github.com/ansible-semaphore/semaphore/api/helpers/helpers.go:92 +0x199
ansible_ui   | github.com/ansible-semaphore/semaphore/api/projects.AddKey({0x2345ac0, 0xc000510000}, 0xc0020e0438?)
ansible_ui   |  /go/src/github.com/ansible-semaphore/semaphore/api/projects/keys.go:91 +0x699
ansible_ui   | net/http.HandlerFunc.ServeHTTP(0x414205?, {0x2345ac0?, 0xc000510000?}, 0x95ee5c?)
ansible_ui   |  /usr/local/go/src/net/http/server.go:2122 +0x2f
ansible_ui   | github.com/ansible-semaphore/semaphore/api/projects.GetMustCanMiddleware.func1.1({0x2345ac0, 0xc000510000}, 0xc00014d800)
ansible_ui   |  /go/src/github.com/ansible-semaphore/semaphore/api/projects/project.go:59 +0x149
ansible_ui   | net/http.HandlerFunc.ServeHTTP(0xc00014d800?, {0x2345ac0?, 0xc000510000?}, 0xbd2660?)
ansible_ui   |  /usr/local/go/src/net/http/server.go:2122 +0x2f
ansible_ui   | github.com/ansible-semaphore/semaphore/api/projects.ProjectMiddleware.func1({0x2345ac0, 0xc000510000}, 0x1?)
ansible_ui   |  /go/src/github.com/ansible-semaphore/semaphore/api/projects/project.go:43 +0x394
ansible_ui   | net/http.HandlerFunc.ServeHTTP(0x2345ac0?, {0x2345ac0?, 0xc000510000?}, 0xd800000000000010?)
ansible_ui   |  /usr/local/go/src/net/http/server.go:2122 +0x2f
ansible_ui   | github.com/ansible-semaphore/semaphore/api.authentication.func1({0x2345ac0, 0xc000510000}, 0xc0005bc000?)
ansible_ui   |  /go/src/github.com/ansible-semaphore/semaphore/api/auth.go:102 +0x5d
ansible_ui   | net/http.HandlerFunc.ServeHTTP(0xbba0c0?, {0x2345ac0?, 0xc000510000?}, 0xc?)
ansible_ui   |  /usr/local/go/src/net/http/server.go:2122 +0x2f
ansible_ui   | github.com/ansible-semaphore/semaphore/api.JSONMiddleware.func1({0x2345ac0, 0xc000510000}, 0x450819?)
ansible_ui   |  /go/src/github.com/ansible-semaphore/semaphore/api/router.go:37 +0xf4
ansible_ui   | net/http.HandlerFunc.ServeHTTP(0xb17300?, {0x2345ac0?, 0xc000510000?}, 0x77ffd8?)
ansible_ui   |  /usr/local/go/src/net/http/server.go:2122 +0x2f
ansible_ui   | github.com/ansible-semaphore/semaphore/api.StoreMiddleware.func1.1()
ansible_ui   |  /go/src/github.com/ansible-semaphore/semaphore/api/router.go:28 +0x2e
ansible_ui   | github.com/ansible-semaphore/semaphore/db.StoreSession({0x23507c8, 0xc0000126e0}, {0xc0020d60d0, 0xc}, 0xc00058d700)
ansible_ui   |  /go/src/github.com/ansible-semaphore/semaphore/db/Store.go:354 +0x68
ansible_ui   | github.com/ansible-semaphore/semaphore/api.StoreMiddleware.func1({0x2345ac0, 0xc000510000}, 0xc00014d800)
ansible_ui   |  /go/src/github.com/ansible-semaphore/semaphore/api/router.go:27 +0xf2
ansible_ui   | net/http.HandlerFunc.ServeHTTP(0xc00014d800?, {0x2345ac0?, 0xc000510000?}, 0xbb6a00?)
ansible_ui   |  /usr/local/go/src/net/http/server.go:2122 +0x2f
ansible_ui   | github.com/ansible-semaphore/semaphore/cli/cmd.runService.func1.1({0x2345ac0, 0xc000510000}, 0xc00058d7b0?)
ansible_ui   |  /go/src/github.com/ansible-semaphore/semaphore/cli/cmd/root.go:74 +0xeb
ansible_ui   | net/http.HandlerFunc.ServeHTTP(0xc016c0?, {0x2345ac0?, 0xc000510000?}, 0x7fe05a6cea68?)
ansible_ui   |  /usr/local/go/src/net/http/server.go:2122 +0x2f
ansible_ui   | github.com/gorilla/mux.CORSMethodMiddleware.func1.1({0x2345ac0, 0xc000510000}, 0xc0020dc120?)
ansible_ui   |  /go/src/github.com/ansible-semaphore/semaphore/vendor/github.com/gorilla/mux/middleware.go:51 +0xa3
ansible_ui   | net/http.HandlerFunc.ServeHTTP(0xc00014d700?, {0x2345ac0?, 0xc000510000?}, 0xc00360?)
ansible_ui   |  /usr/local/go/src/net/http/server.go:2122 +0x2f
ansible_ui   | github.com/gorilla/mux.(*Router).ServeHTTP(0xc00028e000, {0x2345ac0, 0xc000510000}, 0xc00014d500)
ansible_ui   |  /go/src/github.com/ansible-semaphore/semaphore/vendor/github.com/gorilla/mux/mux.go:212 +0x202
ansible_ui   | github.com/gorilla/handlers.ProxyHeaders.func1({0x2345ac0, 0xc000510000}, 0xc00014d500)
ansible_ui   |  /go/src/github.com/ansible-semaphore/semaphore/vendor/github.com/gorilla/handlers/proxy_headers.go:59 +0x144
ansible_ui   | net/http.HandlerFunc.ServeHTTP(0x3b?, {0x2345ac0?, 0xc000510000?}, 0x4e1a66?)
ansible_ui   |  /usr/local/go/src/net/http/server.go:2122 +0x2f
ansible_ui   | github.com/ansible-semaphore/semaphore/cli/cmd.cropTrailingSlashMiddleware.func1({0x2345ac0?, 0xc000510000?}, 0xc00014d500?)
ansible_ui   |  /go/src/github.com/ansible-semaphore/semaphore/cli/cmd/server.go:27 +0xf7
ansible_ui   | net/http.HandlerFunc.ServeHTTP(0x0?, {0x2345ac0?, 0xc000510000?}, 0x725074?)
ansible_ui   |  /usr/local/go/src/net/http/server.go:2122 +0x2f
ansible_ui   | net/http.serverHandler.ServeHTTP({0x23436b0?}, {0x2345ac0, 0xc000510000}, 0xc00014d500)
ansible_ui   |  /usr/local/go/src/net/http/server.go:2936 +0x316
ansible_ui   | net/http.(*conn).serve(0xc000624e10, {0x2346198, 0xc000629f80})
ansible_ui   |  /usr/local/go/src/net/http/server.go:1995 +0x612
ansible_ui   | created by net/http.(*Server).Serve
ansible_ui   |  /usr/local/go/src/net/http/server.go:3089 +0x5ed
```

**Comment by lechl1** [comment]
I have this problem too. Downgrading to previous minor version v2.8.95 helped.

**Comment by koffiisen** [comment]
Thanks you for your reply, but i followed your instructions and i have same problem. Wich database you used ?
For my part i use postgres 13

**Comment by automation4you** [comment]
I just updated to v2.9.45 and have no issues adding new SSH key. 

installed via package manager, using mysql

**Comment by isaacblum** [comment]
I'm having the same problem. I tried a new database on both v2.8.95 and the latest version of the Docker image, and I cannot add it. I can add a non-key, but neither of the SSH keys(private or password based) is working. I'm using the Docker container for both of these tests.

My database is PostgreSQL.

Error from v2.8.95
time="2024-03-19T20:09:22Z" level=error msg="illegal base64 data at input byte 14"
goroutine 55 [running]:
runtime/debug.Stack()
	/usr/local/go/src/runtime/debug/stack.go:24 +0x65
runtime/debug.PrintStack()
	/usr/local/go/src/runtime/debug/stack.go:16 +0x19
github.com/ansible-semaphore/semaphore/api/helpers.WriteError({0x22d26a0, 0xc0007847e0}, {0x22ce360?, 0x26e2570?})
	/go/src/github.com/ansible-semaphore/semaphore/api/helpers/helpers.go:92 +0x1ae
github.com/ansible-semaphore/semaphore/api/projects.AddKey({0x22d26a0, 0xc0007847e0}, 0x22ca460?)
	/go/src/github.com/ansible-semaphore/semaphore/api/projects/keys.go:91 +0x66d
net/http.HandlerFunc.ServeHTTP(0xc001f8d358?, {0x22d26a0?, 0xc0007847e0?}, 0xc000c6a5d0?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/ansible-semaphore/semaphore/api/projects.GetMustCanMiddleware.func1.1({0x22d26a0, 0xc0007847e0}, 0xc000c3ad00)
	/go/src/github.com/ansible-semaphore/semaphore/api/projects/project.go:59 +0x149
net/http.HandlerFunc.ServeHTTP(0xc000c3ad00?, {0x22d26a0?, 0xc0007847e0?}, 0xba3b00?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/ansible-semaphore/semaphore/api/projects.ProjectMiddleware.func1({0x22d26a0, 0xc0007847e0}, 0x0?)
	/go/src/github.com/ansible-semaphore/semaphore/api/projects/project.go:43 +0x455
net/http.HandlerFunc.ServeHTTP(0x22d26a0?, {0x22d26a0?, 0xc0007847e0?}, 0xb70000c00005c400?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/ansible-semaphore/semaphore/api.authentication.func1({0x22d26a0, 0xc0007847e0}, 0xc0008fec60?)
	/go/src/github.com/ansible-semaphore/semaphore/api/auth.go:102 +0x5d
net/http.HandlerFunc.ServeHTTP(0xb8c540?, {0x22d26a0?, 0xc0007847e0?}, 0xc?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/ansible-semaphore/semaphore/api.JSONMiddleware.func1({0x22d26a0, 0xc0007847e0}, 0x1b?)
	/go/src/github.com/ansible-semaphore/semaphore/api/router.go:35 +0xf4
net/http.HandlerFunc.ServeHTTP(0xc001f8d670?, {0x22d26a0?, 0xc0007847e0?}, 0xc00009d700?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/ansible-semaphore/semaphore/api.StoreMiddleware.func1.1()
	/go/src/github.com/ansible-semaphore/semaphore/api/router.go:26 +0x2e
github.com/ansible-semaphore/semaphore/db.StoreSession({0x22dbf30, 0xc0000126d8}, {0xc00009d700, 0x1b}, 0xc001f8d6c8)
	/go/src/github.com/ansible-semaphore/semaphore/db/Store.go:352 +0x68
github.com/ansible-semaphore/semaphore/api.StoreMiddleware.func1({0x22d26a0, 0xc0007847e0}, 0xc000c3ad00)
	/go/src/github.com/ansible-semaphore/semaphore/api/router.go:25 +0xfc
net/http.HandlerFunc.ServeHTTP(0xc000c3ad00?, {0x22d26a0?, 0xc0007847e0?}, 0xb65360?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/ansible-semaphore/semaphore/cli/cmd.runService.func1.1({0x22d26a0, 0xc0007847e0}, 0xc001f8d7a8?)
	/go/src/github.com/ansible-semaphore/semaphore/cli/cmd/root.go:67 +0x124
net/http.HandlerFunc.ServeHTTP(0x7f66716fb128?, {0x22d26a0?, 0xc0007847e0?}, 0x100?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/gorilla/mux.CORSMethodMiddleware.func1.1({0x22d26a0, 0xc0007847e0}, 0xc000c6a510?)
	/go/src/github.com/ansible-semaphore/semaphore/vendor/github.com/gorilla/mux/middleware.go:51 +0xaa
net/http.HandlerFunc.ServeHTTP(0xc000c3ac00?, {0x22d26a0?, 0xc0007847e0?}, 0xbd0740?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/gorilla/mux.(*Router).ServeHTTP(0xc0002a2000, {0x22d26a0, 0xc0007847e0}, 0xc000c3ab00)
	/go/src/github.com/ansible-semaphore/semaphore/vendor/github.com/gorilla/mux/mux.go:212 +0x202
github.com/gorilla/handlers.ProxyHeaders.func1({0x22d26a0, 0xc0007847e0}, 0xc000c3ab00)
	/go/src/github.com/ansible-semaphore/semaphore/vendor/github.com/gorilla/handlers/proxy_headers.go:59 +0x144
net/http.HandlerFunc.ServeHTTP(0x3b?, {0x22d26a0?, 0xc0007847e0?}, 0x72?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/ansible-semaphore/semaphore/cli/cmd.cropTrailingSlashMiddleware.func1({0x22d26a0?, 0xc0007847e0?}, 0xc000c3ab00?)
	/go/src/github.com/ansible-semaphore/semaphore/cli/cmd/server.go:27 +0xf7
net/http.HandlerFunc.ServeHTTP(0x0?, {0x22d26a0?, 0xc0007847e0?}, 0x6ee8b4?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
net/http.serverHandler.ServeHTTP({0x22d04a0?}, {0x22d26a0, 0xc0007847e0}, 0xc000c3ab00)
	/usr/local/go/src/net/http/server.go:2947 +0x30c
net/http.(*conn).serve(0xc0007a4140, {0x22d2fb8, 0xc00078b350})
	/usr/local/go/src/net/http/server.go:1991 +0x607
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:3102 +0x4db
	
	
Error from latest 
time="2024-03-19T21:02:37Z" level=error msg="illegal base64 data at input byte 14"
goroutine 35 [running]:
runtime/debug.Stack()
	/usr/local/go/src/runtime/debug/stack.go:24 +0x65
runtime/debug.PrintStack()
	/usr/local/go/src/runtime/debug/stack.go:16 +0x19
github.com/ansible-semaphore/semaphore/api/helpers.WriteError({0x2345ac0, 0xc00203c1c0}, {0x2341600?, 0x27788b0?})
	/go/src/github.com/ansible-semaphore/semaphore/api/helpers/helpers.go:92 +0x199
github.com/ansible-semaphore/semaphore/api/projects.AddKey({0x2345ac0, 0xc00203c1c0}, 0xc0004e6798?)
	/go/src/github.com/ansible-semaphore/semaphore/api/projects/keys.go:91 +0x699
net/http.HandlerFunc.ServeHTTP(0x414205?, {0x2345ac0?, 0xc00203c1c0?}, 0x95ee5c?)
	/usr/local/go/src/net/http/server.go:2122 +0x2f
github.com/ansible-semaphore/semaphore/api/projects.GetMustCanMiddleware.func1.1({0x2345ac0, 0xc00203c1c0}, 0xc000106500)
	/go/src/github.com/ansible-semaphore/semaphore/api/projects/project.go:59 +0x149
net/http.HandlerFunc.ServeHTTP(0xc000106500?, {0x2345ac0?, 0xc00203c1c0?}, 0xbd2660?)
	/usr/local/go/src/net/http/server.go:2122 +0x2f
github.com/ansible-semaphore/semaphore/api/projects.ProjectMiddleware.func1({0x2345ac0, 0xc00203c1c0}, 0x1?)
	/go/src/github.com/ansible-semaphore/semaphore/api/projects/project.go:43 +0x394
net/http.HandlerFunc.ServeHTTP(0x2345ac0?, {0x2345ac0?, 0xc00203c1c0?}, 0x5c00000000000010?)
	/usr/local/go/src/net/http/server.go:2122 +0x2f
github.com/ansible-semaphore/semaphore/api.authentication.func1({0x2345ac0, 0xc00203c1c0}, 0xc00016a2c0?)
	/go/src/github.com/ansible-semaphore/semaphore/api/auth.go:102 +0x5d
net/http.HandlerFunc.ServeHTTP(0xbba0c0?, {0x2345ac0?, 0xc00203c1c0?}, 0xc?)
	/usr/local/go/src/net/http/server.go:2122 +0x2f
github.com/ansible-semaphore/semaphore/api.JSONMiddleware.func1({0x2345ac0, 0xc00203c1c0}, 0x450819?)
	/go/src/github.com/ansible-semaphore/semaphore/api/router.go:37 +0xf4
net/http.HandlerFunc.ServeHTTP(0xb17300?, {0x2345ac0?, 0xc00203c1c0?}, 0x77ffd8?)
	/usr/local/go/src/net/http/server.go:2122 +0x2f
github.com/ansible-semaphore/semaphore/api.StoreMiddleware.func1.1()
	/go/src/github.com/ansible-semaphore/semaphore/api/router.go:28 +0x2e
github.com/ansible-semaphore/semaphore/db.StoreSession({0x23507c8, 0xc0000126e0}, {0xc0004453a0, 0xc}, 0xc0000d3700)
	/go/src/github.com/ansible-semaphore/semaphore/db/Store.go:354 +0x68
github.com/ansible-semaphore/semaphore/api.StoreMiddleware.func1({0x2345ac0, 0xc00203c1c0}, 0xc000106500)
	/go/src/github.com/ansible-semaphore/semaphore/api/router.go:27 +0xf2
net/http.HandlerFunc.ServeHTTP(0xc000106500?, {0x2345ac0?, 0xc00203c1c0?}, 0xbb6a00?)
	/usr/local/go/src/net/http/server.go:2122 +0x2f
github.com/ansible-semaphore/semaphore/cli/cmd.runService.func1.1({0x2345ac0, 0xc00203c1c0}, 0xc0000d37b0?)
	/go/src/github.com/ansible-semaphore/semaphore/cli/cmd/root.go:74 +0xeb
net/http.HandlerFunc.ServeHTTP(0xc016c0?, {0x2345ac0?, 0xc00203c1c0?}, 0x7f5e6ab3f878?)
	/usr/local/go/src/net/http/server.go:2122 +0x2f
github.com/gorilla/mux.CORSMethodMiddleware.func1.1({0x2345ac0, 0xc00203c1c0}, 0xc00065d4d0?)
	/go/src/github.com/ansible-semaphore/semaphore/vendor/github.com/gorilla/mux/middleware.go:51 +0xa3
net/http.HandlerFunc.ServeHTTP(0xc000106400?, {0x2345ac0?, 0xc00203c1c0?}, 0xc00360?)
	/usr/local/go/src/net/http/server.go:2122 +0x2f
github.com/gorilla/mux.(*Router).ServeHTTP(0xc00029e000, {0x2345ac0, 0xc00203c1c0}, 0xc000106300)
	/go/src/github.com/ansible-semaphore/semaphore/vendor/github.com/gorilla/mux/mux.go:212 +0x202
github.com/gorilla/handlers.ProxyHeaders.func1({0x2345ac0, 0xc00203c1c0}, 0xc000106300)
	/go/src/github.com/ansible-semaphore/semaphore/vendor/github.com/gorilla/handlers/proxy_headers.go:59 +0x144
net/http.HandlerFunc.ServeHTTP(0x3b?, {0x2345ac0?, 0xc00203c1c0?}, 0x4e1a66?)
	/usr/local/go/src/net/http/server.go:2122 +0x2f
github.com/ansible-semaphore/semaphore/cli/cmd.cropTrailingSlashMiddleware.func1({0x2345ac0?, 0xc00203c1c0?}, 0xc000106300?)
	/go/src/github.com/ansible-semaphore/semaphore/cli/cmd/server.go:27 +0xf7
net/http.HandlerFunc.ServeHTTP(0x0?, {0x2345ac0?, 0xc00203c1c0?}, 0x725074?)
	/usr/local/go/src/net/http/server.go:2122 +0x2f
net/http.serverHandler.ServeHTTP({0x23436b0?}, {0x2345ac0, 0xc00203c1c0}, 0xc000106300)
	/usr/local/go/src/net/http/server.go:2936 +0x316
net/http.(*conn).serve(0xc000144240, {0x2346198, 0xc000843f50})
	/usr/local/go/src/net/http/server.go:1995 +0x612
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:3089 +0x5ed

**Comment by isaacblum** [comment]
I can confirm that when using the Docker container on the latest version, but with the MySQL database, the issue doesn't exist.

**Comment by isaacblum** [comment]
Sorry for all the extra messages on this... This was entirely my issue. I missed this part in the documentation: https://docs.semui.co/administration-guide/installation#docker

"SEMAPHORE_ACCESS_KEY_ENCRYPTION — key for encrypting access keys in the database. It must be generated by using the following command: head -c32 /dev/urandom | base64."

Once I updated it with the correct base64 key, I no longer had the issue. Sorry again...

Confirming that the latest Docker image and an external PostgreSQL database are working.
